### PR TITLE
Trading trait and deal with metadata in Mutate trait for nonfungibles_v2

### PR DESCRIPTION
**Pull request by AlexD10S**
I have added some Traits that are missing and are useful for dealing with non-fungible tokens on other pallets and their implementations for NFTs pallet.

- In the Mutate trait, added methods for dealing with the metadata: `set_metadata`,  `set_collection_metadata`, `clear_metadata` and `clear_collection_metadata`.
The motivation of adding this methods coming from a StackExchange question asking for it: [Setting metadata of an item of the Nfts pallet in a custom pallet](https://substrate.stackexchange.com/questions/9974/setting-metadata-of-an-item-of-the-nfts-pallet-in-a-custom-pallet)

- A Trait for trading non-fungible items. The methods in that Trait are `buy_item`, `set_price` and  `item_price`
An example of where this Trait can be useful is a pallet that deals with [NFT Royalties](https://forum.polkadot.network/t/nfts-royalty-pallet/3766) and needs to perform this actions.

### CODE REVIEW

**Review by sacha-l — COMMENTED** [review]
Nice ! 🚀

File: substrate/frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 1]
```diff
@@ -370,3 +414,27 @@ pub trait Transfer<AccountId>: Inspect<AccountId> {
...
+	/// Allows `buyer` to buy an `item` of `collection` if it's up for sale with a `bid_price` to
+	/// pay.
+	fn buy_item(
+		collection: &Self::CollectionId,
+		item: &Self::ItemId,
+		buyer: &AccountId,
+		bid_price: &ItemPrice,
+	) -> DispatchResult;
+
+	/// Set the item price of `item` of `collection` to make it available for sale
```
[the remark this answers is not in the document]
Comment (AlexD10S) [reply substrate/frame/support/src/traits/tokens/nonfungibles_v2.rs]: Done!

File: substrate/frame/nfts/src/impl_nonfungibles.rs [hunk 2]
```diff
@@ -320,6 +320,33 @@ impl<T: Config<I>, I: 'static> Mutate<<T as SystemConfig>::AccountId, ItemConfig
 		})
 	}
 
+	fn set_metadata(
```
Comment (jsidorenko) [inline substrate/frame/nfts/src/impl_nonfungibles.rs:323]: shall we name it as `set_item_metadata`, wdyt?

File: substrate/frame/nfts/src/impl_nonfungibles.rs [hunk 3]
```diff
@@ -320,6 +320,33 @@ impl<T: Config<I>, I: 'static> Mutate<<T as SystemConfig>::AccountId, ItemConfig
...
+		who: &T::AccountId,
```
Comment (jsidorenko) [inline substrate/frame/nfts/src/impl_nonfungibles.rs:324]: I think we should remove that `who` param and follow the same pattern we set the attributes by passing `None` into the `do_*` method

Comment (jsidorenko) [reply substrate/frame/nfts/src/impl_nonfungibles.rs:324]: Or we can make it optional :) maybe this would even be a more universal approach

Comment (jsidorenko) [reply substrate/frame/nfts/src/impl_nonfungibles.rs:324]:
```
	fn set_metadata(
		who: Option<&T::AccountId>,
		collection: &Self::CollectionId,
		item: &Self::ItemId,
		data: &[u8],
	) -> DispatchResult {
		Self::do_set_item_metadata(
			who.clone(),
			*collection,
			*item,
			Self::construct_metadata(data.to_vec())?,
			None,
		)
	}
```

File: substrate/frame/nfts/src/impl_nonfungibles.rs [hunk 4]
```diff
@@ -320,6 +320,33 @@ impl<T: Config<I>, I: 'static> Mutate<<T as SystemConfig>::AccountId, ItemConfig
...
+			Some(who.clone()),
+			*collection,
+			*item,
+			Self::construct_metadata(data.to_vec())?,
+			None,
+		)
+	}
+
+	fn set_collection_metadata(
+		who: &T::AccountId,
```
Comment (jsidorenko) [inline substrate/frame/nfts/src/impl_nonfungibles.rs:339]: Same as above:
```suggestion
		who: Option<&T::AccountId>,
```

File: substrate/frame/nfts/src/impl_nonfungibles.rs [hunk 5]
```diff
@@ -362,6 +389,21 @@ impl<T: Config<I>, I: 'static> Mutate<<T as SystemConfig>::AccountId, ItemConfig
 			<Self as Mutate<T::AccountId, ItemConfig>>::clear_collection_attribute(collection, k)
 		})
 	}
+
+	fn clear_metadata(
+		who: &T::AccountId,
```
Comment (jsidorenko) [inline substrate/frame/nfts/src/impl_nonfungibles.rs:394]:
```suggestion
		who: Option<&T::AccountId>,
```

File: substrate/frame/nfts/src/impl_nonfungibles.rs [hunk 6]
```diff
@@ -362,6 +389,21 @@ impl<T: Config<I>, I: 'static> Mutate<<T as SystemConfig>::AccountId, ItemConfig
...
+	) -> DispatchResult {
+		Self::do_clear_item_metadata(Some(who.clone()), *collection, *item)
+	}
+
+	fn clear_collection_metadata(
+		who: &T::AccountId,
+		collection: &Self::CollectionId,
+	) -> DispatchResult {
+		Self::do_clear_collection_metadata(Some(who.clone()), *collection)
+	}
```
Comment (jsidorenko) [inline substrate/frame/nfts/src/impl_nonfungibles.rs:406]:
```suggestion
	fn clear_collection_metadata(
		who: Option<&T::AccountId>,
		collection: &Self::CollectionId,
	) -> DispatchResult {
		Self::do_clear_collection_metadata(who.clone(), *collection)
	}
```

File: substrate/frame/nfts/src/impl_nonfungibles.rs (hunk 2 above)
Comment (AlexD10S) [reply substrate/frame/nfts/src/impl_nonfungibles.rs]: Is calling `do_set_item_metadata` so for me makes sense to change the naming and use `set_item_metadata`

File: substrate/frame/nfts/src/impl_nonfungibles.rs (hunk 3 above)
Comment (AlexD10S) [reply substrate/frame/nfts/src/impl_nonfungibles.rs]: Done. I used the optional approach

File: substrate/frame/nfts/src/impl_nonfungibles.rs (hunk 5 above)
Comment (AlexD10S) [reply substrate/frame/nfts/src/impl_nonfungibles.rs]: Done! Also changed the naming to `clear_item_metadata` as we did for the one above

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3981653

**Review by jsidorenko — APPROVED** [review]

**Review by kianenigma — APPROVED** [review]

**Review by seadanda — APPROVED** [review]

**Comment by jsidorenko** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-3-0/4614/1

**Pull request merged**